pgpainless_signer/src/main/java/org/eclipse/packager/rpm/signature/pgpainless/PGPainlessSignatureProcessor.java

```diff
@@ -0,0 +1,122 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+
+public class PGPainlessSignatureProcessor implements SignatureProcessor {
+
+    protected final EncryptionStream signingStream;
+    private final Logger logger = LoggerFactory.getLogger(PGPainlessSignatureProcessor.class);
+
+    public PGPainlessSignatureProcessor(PGPSecretKeyRing key, SecretKeyRingProtector keyProtector, int hashAlgorithm) {
+        OutputStream sink = new OutputStream() {
+            @Override
+            public void write(int i) {
+                // Discard plaintext
+            }
+        };
+        try {
+            signingStream = PGPainless.encryptAndOrSign()
+                .onOutputStream(sink)
+                .withOptions(
+                    ProducerOptions.sign(
+                        SigningOptions.get()
+                            .overrideHashAlgorithm(HashAlgorithm.requireFromId(hashAlgorithm))
+                            .addDetachedSignature(keyProtector, key)
+                    ).setAsciiArmor(false)
+                        .overrideCompressionAlgorithm(CompressionAlgorithm.UNCOMPRESSED)
+                );
+        } catch (PGPException | IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void feedHeader(ByteBuffer header) {
+        feedData(header);
+    }
+
+    @Override
+    public void feedPayloadData(ByteBuffer data) {
+        feedData(data);
+    }
+
+    private void feedData(ByteBuffer data) {
+        try {
+            if (data.hasArray()) {
+                signingStream.write(data.array(), data.position(), data.remaining());
+            } else {
+                final byte[] buffer = new byte[data.remaining()];
+                data.get(buffer);
+                signingStream.write(buffer);
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    protected Logger getLogger() {
+        return logger;
+    }
+
+    @Override
+    public void finish(Header<RpmSignatureTag> signature) {
+        try {
+            signingStream.close();
+            EncryptionResult result = signingStream.getResult();
+            PGPSignature pgpSignature = result.getDetachedSignatures().flatten().iterator().next();
+            byte[] value = pgpSignature.getEncoded();
+            switch (pgpSignature.getKeyAlgorithm()) {
+                // RSA
+                case PublicKeyAlgorithmTags.RSA_GENERAL: // 1
+                    getLogger().info("RSA: {}", value);
+                    signature.putBlob(RpmSignatureTag.PGP, value);
+                    break;
+
+                // DSA
+                case PublicKeyAlgorithmTags.DSA: // 17
+                case PublicKeyAlgorithmTags.EDDSA_LEGACY: // 22
+                    getLogger().info("DSA: {}", value);
+                    signature.putBlob(RpmSignatureTag.GPG, value);
```

Comment:
`RpmSignatureTag.GPG` was missing, so I added it.
Do I understand the protocol correctly in that `RsaSignatureProcessor` (and analogue `PGPainlessSignatureProcessor`) do implement the "RPM v3 signature with the same key on the header+payload" the manual is talking about?

Again, the documentation only mentions ECDSA, not EdDSA.

Reply:
We support RPMv4, not RPMv3, so I don't think we want the older tags mentioned there.

Reply:
I mean, you can keep the constants defined, but I am not sure we want to write them in the header.

Reply:
RsaSignatureProcessor is using RpmSignatureTag.PGP [here](https://github.com/eclipse/packager/blob/master/rpm/src/main/java/org/eclipse/packager/rpm/signature/RsaSignatureProcessor.java#L84).

Reply:
> RsaSignatureProcessor is using RpmSignatureTag.PGP (which as far as I understand is RPM v3) [here](https://github.com/eclipse/packager/blob/master/rpm/src/main/java/org/eclipse/packager/rpm/signature/RsaSignatureProcessor.java#L84).

It depends what version of RPM we are emulating. I can find "RPMSIGTAG_PGP, RPMSIGTAG_GPG (and RPMSIGTAG_PGP5) are no longer created in >= 4.16". So, we can have both sets, but if we are going to only have one set, it should not be these.

I think the new ones only process the header, whereas the old ones processed the header together with the payload.

Reply:
I'm still confused.
To slim down the PR I'd remove the new constant again.
However, what `RpmSignatureTag` should I use in the `PGPainlessSignatureProcessor` for EdDSA/DSA signatures?
If `RpmSignatureTag.PGP` (used for RSA) is also no longer created in newer Rpm versions, shall I just remove the `PGPainlessSignatureProcessor` altogether? What about its BC counterpart (`RsaSignatureProcessor`)? Remove that as well? It appears it is still in use in `RpmFileSignatureProcessor`. Is this class also obsolete then?

Reply:
There are two signatures. One is header-only and one is header and payload. That's 2 possible for RSA and two possible for DSA/EdDSA making a total of 4 signatures.

Reply:
So RpmFileSignatureProcessor currently uses RPMv3 signing then (RsaSignatureProcessor uses PGP tag).
I guess I'll leave RsaSignatureProcessor untouched and also won't change RpmFileSignatureProcessor in this PR.

I'll remove PGPainlessSignatureProcessor (which would do RPMv3 signing) and instead only implement PGPainlessHeaderSignatureProcessor.